### ISSUE TITLE
Provide interfaces for all services

### ISSUE THREAD

**Issue by nozzlegear** [nozzlegear/ShopifySharp#362]
I've been using a lot of dependency injection lately, but have found it mildly difficult to test any function that uses a ShopifySharp service since the package provides no interfaces for the services.

**Comment by evaldas-raisutis** [comment]
I've been lately using ShopifySharp (cool lib!) in a DI heavy project which also runs integration tasks in multi tenant environment. Having interfaces would help a great deal, but I've also been slightly struggling with multi-tenant aspect of our integrations.

I ended up solving it so far with a two-fold solution.

Firstly, all resource services are provided by injecting a factory method into a constructor (for a command handler, for example) like:

```
public class CreateProductHandler : ICommandHandler<CreateProduct>
{
    private readonly Func<ShopAlias, ProductService> productServiceFactory;

    public CreateProductHandler(Func<ShopAlias, ProductService> productServiceFactory)
    {
        this.productServiceFactory = productServiceFactory;
    }
    public Task Handle(CreateProduct command)
    {
        var productService = this.productServiceFactory.Invoke(command.shopAlias);
        // ...
    }
}
```

This allows to get tenant-appropriate instance of ProductService by invoking the factory method. ShopAlias is passed down through a Command message. DI then has to be setup at Startup of course, for example with IServiceCollection:

services.Add<Func<ShopAlias, ProductService>>(x => shopAlias => 
{
  var shopConfigurations = x.GetService<ShopConfigurations>();
  var shopConfiguration = shopConfiguration.ByAlias(shopAlias);

  return new ProductService(shopConfiguration.Url, shopConfiguration.Token);
});

Now back to topic at hand, interfaces on Services would be great, since I could replace 'new' part with something that calls a factory service instead and return any implementation of IProductService.

**Comment by Alfetta159** [comment]
I find that DI is not necessarily an issue, but unit testing is painful w/o interfaces on all of the non-data object classes.

**Comment by nozzlegear** [comment]
I agree that testing can be annoying without interfaces. I'm looking into implementing interfaces for all of the service classes within the next couple of weeks. Hopefully there's some kind of Roslyn analyzer tool out there that can do it all for me and save me a lot of copying/pasting! 😅

**Comment by nozzlegear** [comment]
Interfaces are being implemented in this PR: #891. Merging soon!

**Comment by nozzlegear** [comment]
All services except for two – AuthorizationService and MultipassService, because they're static – now have interfaces as of ShopifySharp 6.1.1 on Nuget.
